### ISSUE TITLE
RemoveProvider in ILoggerFactory

### ISSUE THREAD

**Issue by prasannavl** [aspnet/Logging#300]
Is there a way to dynamically remove a log provider? I wonder why it wasn't a part of the initial design decision. Unless, I'm missing something, it should've been the natural course of design. 

This will help immensely to add and remove loggers in runtime based on conditions. Say when need, a provider could be add to enable deep logging on production systems on demand.

**Comment by Eilon** [comment]
@lodejard thoughts?

Though I'm wondering, can providers even be *added* at runtime (that is, after Startup)?

**Comment by lodejard** [comment]
Yep, they can be added at any time. The default ILoggerFactory keeps track of the ILogger instances that have been created in order to attach the new provider on the fly. (that code is thread safe, you can add a provider while existing loggers are being written to) (plus it helps reduce overhead to return the same ILogger every time it's asked for).

Part of the reason removing isn't implemented is because it's very simpler to add the provider's ilogger when the provider is added, but much trickier to keep track of which iloggers came from which provider to remove them correctly.

For this particular case I would suggest implementing an ILoggerProvider wrapper class which takes an `ILoggerProvider Inner` and `bool Enabled` properties. Each call to the wrapper CreateLogger could call the Inner.CreateLogger and return a new WrappingLogger instance. That WrappingLogger implementation would check _parent.Enabled and if it is false it would not call through to the underlying logger.

In a nutshell, any of the existing loggers could be added at startup behind the wrapper, and the wrapper would be so application would have a simple way of toggling them each on or off. 

That said - several logging systems have first-class support for dialing around their settings at runtime. Depending on the system you're using that might actually be the first place to look - to dial the verbosity per category around at runtime in a much more fine-grained fashion.

**Comment by prasannavl** [comment]
@lodejard, ILoggerProvider Wrapper is a solution. But looks like a lot of work for a simple scenario. The major problem I see is in the current internal implementation of the `LoggerFactory` and `Logger`, but not the concept itself.  

Superficially speaking if a `DefaultLoggerProvider` could be created which tracks its own `Loggers` - this problem could be alleviated. Now, the new problem would be to caching the `Loggers` for `CreateLogger` of the `LoggerFactory`, which I think should be separated out to something like a  `LoggerCache` explicitly.  Does this make sense?

**Comment by prasannavl** [comment]
Or, I guess an easier option would be to just create another map that maps ILoggers to their providers.

**Comment by prasannavl** [comment]
Add a quick implementation to `RemoveProvider` only to the core classes, not the interfaces. Let me know if this acceptable. Then we can propagate that to the interfaces, add tests etc.

**Comment by Eilon** [comment]
At this time we do not believe that there are sufficiently compelling scenarios to necessitate "remove" functionality on the logger factory. Ultimately a "real" logger will need to be used in any production application because none of the built-in loggers are persistent.
